### ISSUE TITLE
Migrate sample stacks to composition

### ISSUE THREAD

**Issue by muvaf** [crossplane/crossplane#1586]
### What problem are you facing?

They are still using claims and classes.

### How could Crossplane help solve your problem?

We can migrate them to composition.

**Comment by muvaf** [comment]
Depends on https://github.com/crossplane/crossplane/issues/1587 so that all sample stacks and Wordpress app target the same definitions.

**Comment by muvaf** [comment]
Depends on https://github.com/crossplane/crossplane/issues/1605 so that Wordpress can still schedule to the target cluster.

**Comment by muvaf** [comment]
Here is the list of PRs and issues that are required for Wordpress to work on composition e2e:
* KAR needs to work with just a Secret  - PR https://github.com/crossplane/crossplane/pull/1604
* There is no way to include InfraDef  in a package, which is required if we'd like to have stack define its own API for network resources; the alternative is to dump network resources just like today and only change the classes with Composition s (which is what the earlier commit in stack-gcp-sample does) - Issue https://github.com/crossplane/crossplane/issues/1610
* Composite resource deletion should be blocked until all composed resources are gone - Issue https://github.com/crossplane/crossplane/issues/1612
* Making Wordpress work with new KAR field and use common types: PRs https://github.com/crossplane/app-wordpress/pull/68 and https://github.com/crossplane/app-wordpress/pull/69
* Using composition in stack-gcp-sample (once this works e2e, others will be opened): PR https://github.com/crossplane/stack-gcp-sample/pull/55

There are a few smaller issues as well like https://github.com/crossplane/crossplane/issues/1611 but the ones that are listed above are blockers for the open PRs to be merged. I think we should be able to include them in 0.13 but we haven't yet done the planning for 0.13 yet.

**Comment by muvaf** [comment]
We've synced and here is the current status on that migration.

The biggest hurdle to do this migration is to be able to bundle an `InfrastructureDefinition` in a package that will deploy it [only once](https://github.com/crossplane/crossplane/issues/1610), as opposed per CR which is how template stacks work right now. Essentially, we need a treatment for `InfrastructureDefinition` as if it's a CRD that the package deploys and this will be addressed in [package manager refactor design](https://github.com/crossplane/crossplane/pull/1616).

However, we would like to deprecate templating-controller and replace it with something else in the new design. So, even if we did implement a way to bundle an `InfrastructureDefinition` today with current package manager as an intermediate state to full transition, we'll have to use templating-controller and then when the design gets implemented there will be another change because the interface will change since templating-controller and `StackDefinition` will be replaced. So, the intermediate state will not change the interface but only the underlying mechanics and then we will change the interface later on.

What we'd like to achieve is that these stacks and Wordpress to use composition end-to-end and not include `StackDefinition` and templating-controller at all. So, in order to reduce the churn and not waste engineering time, we will keep the stacks and Wordpress to use claims and classes until we get that new package refactor design landed and confirm that we got a mechanism to achieve what we do with templating-controller.

The plan is that it is expected that the claims&classes will be removed by v0.15 but odds are we'll have the alternative mechanics in place before that and will have a chance to do some manual testing. In the worst case, we can keep claim&classes one more release. The point is that it doesn't really benefit users to change the underlying mechanism without replacing the interface (CRs created by user) that represents the final state we'd like to be in. Although one benefit could be that we'd confirm everything works, i.e. dog-fooding composition, but that's already done; I have taken manual actions to fix the issues listed in the earlier comment and confirmed Wordpress works e2e if those issues are addressed.

Another reason for stopping this work is that `KubernetesApplication` and `KubernetesApplicationResource`, i.e. workload scheduling API, will be deprecated and Wordpress app depends on them very tightly. The alternative model where a Crossplane Agent will be deployed to the remote clusters will replace our deployment model of apps and that's not there yet either. So, there will yet another big churn both in the interface and underlying mechanics of Wordpress app.

In conclusion, we'll keep everything as is and working. The more stuff gets landed we will either open new repos for stacks and WP or have draft PRs that confirm that we're heading in the right direction.

I will leave the migration PR to composition open as draft because this will not need to change much https://github.com/crossplane/stack-gcp-sample/pull/55

**Comment by negz** [comment]
I removed 'Wordpress' from the title thinking it meant to implement wordpress via Composition. I realised right after that this meant moving the Wordpress controller to make requirements, not claims. @muvaf do you think we can close this issue? Based on your summary it seems like we landed at not doing this work - or do you think it's an open question still?

**Comment by muvaf** [comment]
We landed at not doing this work right now until all the missing pieces are there, like being able to have InfraDef in a package. I believe once we do have those things, we'll want to do the migration though it may not be called _stack_ anymore. So, I'd keep it open.

**Comment by hasheddan** [comment]
Feel like this is somewhat addressed by the introduction of `provider-helm`? Do you think we should keep this open @muvaf ?

**Comment by muvaf** [comment]
> Feel like this is somewhat addressed by the introduction of provider-helm

@hasheddan I think the reference platforms are what used to be called sample stacks. If we have an issue about their implementation (or if we don't have a plan to implement them), we can certainly close this.

**Comment by muvaf** [comment]
Fixed by https://github.com/upbound/platform-ref-aws
